keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+---
+kep-number: 17
+title: Pipe Tasks
+authors:
+  - "@zen-dog"
+owners:
+  - "@zen-dog"
+editor: TBD
+creation-date: 2019-09-12
+last-updated: 2019-09-12
+status: provisional
+---
+
+# Pipe Tasks
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Limitations](#limitations)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+* [Alternatives](#alternatives)
+
+## Summary
+
+Developing complicated operators often require generating files in one task and reusing them in a later one. A common example is generating custom certificates/dynamic configuration files in the bootstrap task and using them in the later deployment step of the service. This KEP describes how KUDO could help operator developers by making this task seamless and automated.
+
+## Motivation
+
+We want to ease the development of operators with complicated life-cycles by enabling to pipe files from one task to another. While this can be achieved in a manual way, involving custom scripts, this is exactly the kind of "magic glue" that KUDO could provide to the operator developers to ease the development process.
+
+### Goals
+
+The goal of this KEP is to describe application examples, limitations and ways to implement passing files between tasks. We aim to help operator developers pipe small files (<1Mb) like certificates and dynamically generated configuration which can be referenced later as ConfigMaps and Secrets.
+
+### Non-Goals
+
+Allowing to pipe all kind of files (>1Mb) between tasks requires a general-purpose storage solution (e.g. S3) available in the cluster. While certainly possible, this is not the goal of this KEP.
+
+## Proposal
+
+This section describes how pipe tasks and files they produce are configured in the operator. Here is a task definition that produces a file that will be stored as a Secret, referenced by `{{.Pipes.Certificate}}` key:
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+            - pipe:
+                - file: /usr/share/MyKey.key
+                  kind: Secret # ConfigMap
```

Comment:
Can we mount the file into the pod if it already exists? That way it could read and update a config.

Reply:
I'm not sure I understand the problem. In general, one can also `kubectl cp local-file namespace/pod:/tmp` copy files into a running Pod.  When would we need this?

Reply:
I was thinking if the previous contents of the file were important to generating your output you may want to be able to access the old version. I didn't have a specific use-case in mind though.

Reply:
Ah, got it. So far I was thinking about pipe-tasks as idempotent, side-effect free and able to rerun if necessary. But your case is also interesting: we could inject another init-container (before the user-specified one) that would pull the previous version of the referenced file into the shared volume 🤔
I think, we need a solid use-case for that, but the proposed implementation certainly leaves room for this extension should it become necessary.